pkg/ipcache/cidr.go

```diff
@@ -77,49 +52,82 @@ func allocateCIDRs(prefixes []*net.IPNet, newlyAllocatedIdentities map[string]*i
 		newlyAllocatedIdentities = map[string]*identity.Identity{}
 	}
 
-	for _, prefix := range prefixes {
-		if prefix == nil {
+	allocatedIdentities := make(map[string]*identity.Identity, len(prefixes))
+	for _, p := range prefixes {
+		if p == nil {
 			continue
 		}
 
-		prefixStr := prefix.String()
-
-		// Figure out if this call needs to be able to update the selector cache synchronously.
-		allocateCtx, cancel := context.WithTimeout(context.Background(), option.Config.IPAllocationTimeout)
-		defer cancel()
+		prefixStr := p.String()
 
-		if IdentityAllocator == nil {
-			return nil, fmt.Errorf("IdentityAllocator not initialized!")
-		}
-
-		id, isNew, err := IdentityAllocator.AllocateIdentity(allocateCtx, cidr.GetCIDRLabels(prefix), false)
+		id, isNew, err := allocate(prefixStr, cidr.GetCIDRLabels(p))
 		if err != nil {
 			IdentityAllocator.ReleaseSlice(context.Background(), nil, usedIdentities)
-			return nil, fmt.Errorf("failed to allocate identity for cidr %s: %s", prefixStr, err)
+			return nil, err
 		}
 
-		id.CIDRLabel = labels.NewLabelsFromModel([]string{labels.LabelSourceCIDR + ":" + prefixStr})
-
 		usedIdentities = append(usedIdentities, id)
 		allocatedIdentities[prefixStr] = id
 		if isNew {
 			newlyAllocatedIdentities[prefixStr] = id
 		}
 	}
 
-	allocatedIdentitiesSlice := make([]*identity.Identity, 0, len(allocatedIdentities))
-
 	// Only upsert into ipcache if identity wasn't allocated
 	// before and the caller does not care doing this
 	if upsert {
 		UpsertGeneratedIdentities(newlyAllocatedIdentities)
 	}
 
+	identities := make([]*identity.Identity, 0, len(allocatedIdentities))
 	for _, id := range allocatedIdentities {
-		allocatedIdentitiesSlice = append(allocatedIdentitiesSlice, id)
+		identities = append(identities, id)
+	}
+	return identities, nil
+}
+
+// AllocateCIDRsForIPs attempts to allocate identities for a list of CIDRs. If
```

Comment:
Not introduced by this PR but maybe worth fixing:
```suggestion
// AllocateCIDRsForIPs attempts to allocate identities for a list of IP addresses. If
```
Or might be best to just say:
```go
// AllocateCIDRsForIPs performs the same action as AllocateCIDRs but for IP addresses instead of CIDRs.
```